### PR TITLE
Resto druid - Added azerite trait Early Harvest 

### PR DESCRIPTION
**Pull request by Qbz23**
image

### CODE REVIEW

File: src/parser/druid/restoration/modules/items/azeritetraits/EarlyHarvest.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    this.active = this.selectedCombatant.hasTrait(SPELLS.EARLY_HARVEST_TRAIT.id);
+  }
+
+  // Save innervate timestamp/cooldown to check it's actually on cd when cd would be reduced by early harvest 
+  on_byPlayer_cast(event) {
+    if(SPELLS.INNERVATE.id === event.ability.guid && event.sourceId === event.targetId) {
+      this.currentInnervateCooldown = INNERVATE_COOLDOWN + INNERVATE_DURATION;
+      this.lastInnervateTimestamp = event.timestamp;
+    }
+  }
```
Comment (Pewtro) [inline src/parser/druid/restoration/modules/items/azeritetraits/EarlyHarvest.js]: SpellUsable can handle this.

File: src/parser/druid/restoration/modules/items/azeritetraits/EarlyHarvest.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    // Check if innervate is on cd 
+    const timeSinceLastInnervate = event.timestamp - this.lastInnervateTimestamp;
+    const innervateCooldownRemaining = this.currentInnervateCooldown - timeSinceLastInnervate;
+    if(innervateCooldownRemaining > 0) {
+        this.currentInnervateCooldown -= COOLDOWN_REDUCTION;
+        // min to handle if on cooldown with less than COOLDOWN_REDUCTION milliseconds remaining 
+        this.innervateCooldownReduced += Math.min(COOLDOWN_REDUCTION, innervateCooldownRemaining);
+        this.numEarlyHarvestCooldownReductions += 1;
+    }
+  }
```
Comment (Pewtro) [inline src/parser/druid/restoration/modules/items/azeritetraits/EarlyHarvest.js]: SpellUsable can handle this aswell

File: src/parser/druid/restoration/modules/items/azeritetraits/EarlyHarvest.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  statistic(){
+    const healingFromWildGrowthExpiration = this.owner.getPercentageOfTotalHealingDone(this.healingFromWildGrowthExpiration);
+    const innervateCooldownReducedInSeconds = this.innervateCooldownReduced/1000.0;
+    
+    return(
+      <TraitStatisticBox
+        position={STATISTIC_ORDER.OPTIONAL()}
+        trait={SPELLS.EARLY_HARVEST_TRAIT.id}
+        value={(
+          <>
```
Comment (Pewtro) [inline src/parser/druid/restoration/modules/items/azeritetraits/EarlyHarvest.js]: position prop not necessary unless you use it for something specific when it comes to ordering (which you're not currently it seems)

Comment (Qbz23) [reply src/parser/druid/restoration/modules/items/azeritetraits/EarlyHarvest.js]: Addressed in commit ce6a2ba849b30945139559efb9490f713436f4aa

File: src/parser/druid/restoration/modules/items/azeritetraits/EarlyHarvest.js (hunk 2 above)
Comment (Qbz23) [reply src/parser/druid/restoration/modules/items/azeritetraits/EarlyHarvest.js]: addressed in commit ce6a2ba849b30945139559efb9490f713436f4aa

File: src/parser/druid/restoration/modules/items/azeritetraits/EarlyHarvest.js (hunk 1 above)
Comment (Qbz23) [reply src/parser/druid/restoration/modules/items/azeritetraits/EarlyHarvest.js]: addressed in commit ce6a2ba849b30945139559efb9490f713436f4aa

File: src/parser/druid/restoration/modules/items/azeritetraits/EarlyHarvest.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    }      
+    
+    if(this.spellUsable.isOnCooldown(SPELLS.INNERVATE.id)) {
+      const invCooldownRemaining = this.spellUsable.cooldownRemaining(SPELLS.INNERVATE.id);
+      // min to handle if on cooldown with less than COOLDOWN_REDUCTION milliseconds remaining 
+      const cooldownReduced = Math.min(COOLDOWN_REDUCTION, invCooldownRemaining);
+      this.innervateCooldownReduced += cooldownReduced;
+      this.spellUsable.reduceCooldown(SPELLS.INNERVATE.id, cooldownReduced);
+      this.numEarlyHarvestCooldownReductions += 1;
+    }
```
[the remark this answers is not in the document]
Comment (Qbz23) [reply src/parser/druid/restoration/modules/items/azeritetraits/EarlyHarvest.js]: Not keeping track of wasted potential, but still need it to not over-report how much the cd was reduced. But it looks like I can do that with the return of reduceCooldown anyway. 

Changed in commit 67dc2f4afa5f1b7c8c5dcd10a818e6ecceafecc5

[the remark this answers is not in the document]
Comment (Pewtro) [reply src/parser/druid/restoration/modules/items/azeritetraits/EarlyHarvest.js]: You could spare 1 LoC with a change more, but this is fine as is 😄

**Comment by Pewtro** [comment]
Thanks a lot 👍

**Pull request merged**